File: web/src/api/hid.ts
import { http } from '@/lib/http.ts';

// paste
export function paste(content: string, langue: string) {
  return http.post('/api/hid/paste', { content, langue });
}

// reset hid
export function reset() {
  return http.post('/api/hid/reset');
}

// get hid mode
export function getHidMode() {
  return http.get('/api/hid/mode');
}

// set hid mode
export function setHidMode(mode: string) {
  const data = {
    mode
  };
  return http.post('/api/hid/mode', data);
}

// get shortcuts
export function getShortcuts() {
  return http.get('/api/hid/shortcuts');
}

// add shortcut
export function addShortcut(keys: any[]) {
  const data = {
    keys
  };
  return http.post('/api/hid/shortcut', data);
}

// delete shortcut
export function deleteShortcut(id: string) {
  const data = {
    id
  };
  return http.delete('/api/hid/shortcut', data);
}

// get shortcut leader key
export function getLeaderKey() {
  return http.get('/api/hid/shortcut/leader-key');
}

// set shortcut leader key
export function setLeaderKey(key: string) {
  return http.post('/api/hid/shortcut/leader-key', { key });
}

File: server/router/hid.go
package router

import (
	"github.com/gin-gonic/gin"

	"NanoKVM-Server/authn"
	"NanoKVM-Server/middleware"
	"NanoKVM-Server/service/hid"
)

const internalUSBRecoverPath = "/api/internal/usb/recover"

func HIDLoopbackHTTPAllowedPaths() []string {
	return []string{internalUSBRecoverPath}
}

func hidRouter(r *gin.Engine) {
	service := hid.NewService()
	api := r.Group("/api").Use(middleware.CheckToken())
	localAPI := r.Group("/api/internal").Use(middleware.CheckLoopbackInternalToken())

	api.POST("/hid/paste", service.Paste) // paste

	api.GET("/hid/shortcuts", service.GetShortcuts)           // get shortcuts
	api.GET("/hid/shortcut/leader-key", service.GetLeaderKey) // get shortcut leader key

	api.GET("/hid/mode", service.GetHidMode) // get hid mode
	api.GET("/hid/leds", service.GetKeyboardLedStatus)

	admin := r.Group("/api").Use(
		middleware.CheckToken(),
		middleware.RequireRole(authn.RoleAdmin),
	)
	admin.POST("/hid/mode", service.SetHidMode) // set hid mode
	admin.POST("/hid/reset", service.ResetHid)  // reset hid
	admin.POST("/hid/shortcut", service.AddShortcut)
	admin.DELETE("/hid/shortcut", service.DeleteShortcut)
	admin.POST("/hid/shortcut/leader-key", service.SetLeaderKey)

	localAPI.POST("/usb/recover", service.RecoverUSB)
}
